File: JAVA/14_Backtracking/BOJ_13023_ABCDE.java
import java.io.IOException;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.StringTokenizer;

public class Main {

    static int n, m;
    static boolean[] visited;
    static ArrayList<Integer>[] list;

    public static void dfs(int dpth, int idx) {
        if (dpth == 4) {
            System.out.println(1);
            System.exit(0);
        }
        visited[idx] = true;
        if (list[idx] != null) {
            for (int e : list[idx]) {
                if (!visited[e]) {
                    dfs(dpth + 1, e);
                }
            }
        }
        visited[idx] = false;
    }

    @SuppressWarnings("unchecked")
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine(), " ");
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        visited = new boolean[n];
        list = new ArrayList[n];
        for (int i = 0; i < n; i++) {
            list[i] = new ArrayList<Integer>();
        }
        for (int i = 0; i < m; i++) {
            st = new StringTokenizer(br.readLine(), " ");
            int x = Integer.parseInt(st.nextToken());
            int y = Integer.parseInt(st.nextToken());
            list[x].add(y);
            list[y].add(x);
        }
        for (int i = 0; i < n; i++) {
            dfs(0, i);
        }
        System.out.println(0);
    }
}

File: C/14_Backtracking/BOJ_13023_ABCDE.c
#include <stdio.h>
int n, m, arr[2000][2000], visited[2000], cnt[2000] = { 0 };
void dfs(int dpth, int idx) {
	if (dpth == 4) {
		printf("1");
		exit(0);
	}
	visited[idx] = 1;
	for (int i = 0; i < cnt[idx]; i++) {
		if (!visited[arr[idx][i]]) {
			dfs(dpth + 1, arr[idx][i]);
		}
	}
	visited[idx] = 0;
}
main() {
	scanf("%d %d", &n, &m);
	for (int i = 0; i < m; i++) {
		int x, y;
		scanf("%d %d", &x, &y);
		arr[x][cnt[x]++] = y;
		arr[y][cnt[y]++] = x;
	}
	for (int i = 0; i < n; i++) {
		dfs(0, i);
	}
	printf("0");
}